store/pruning/manager.go

```diff
@@ -1,166 +1,89 @@
-	}
-}
-
-// SetStorageOptions sets the state storage options.
-func (m *Manager) SetStorageOptions(opts Options) {
-	m.mtx.Lock()
-	defer m.mtx.Unlock()
-
-	m.storageOpts = opts
-}
-
-// SetCommitmentOptions sets the state commitment options.
-func (m *Manager) SetCommitmentOptions(opts Options) {
-	m.mtx.Lock()
-	defer m.mtx.Unlock()
-
-	m.commitmentOpts = opts
-}
-
-// Start starts the manager.
-func (m *Manager) Start() {
-	m.mtx.Lock()
-	defer m.mtx.Unlock()
-
-	if m.isStarted {
-		return
-	}
-	m.isStarted = true
-
-	if !m.storageOpts.Sync {
-		m.chStorage = make(chan struct{}, 1)
-		m.chStorage <- struct{}{}
-	}
-	if !m.commitmentOpts.Sync {
-		m.chCommitment = make(chan struct{}, 1)
-		m.chCommitment <- struct{}{}
+		storageOpts:     storageOpts,
+		commitmentOpts:  commitmentOpts,
 	}
 }
 
-// Stop stops the manager and waits for all goroutines to finish.
-func (m *Manager) Stop() {
-	m.mtx.Lock()
-	defer m.mtx.Unlock()
-
-	if !m.isStarted {
-		return
+// Prune prunes the state storage and state commitment.
+// It will check the pruning conditions and prune if necessary.
+func (m *Manager) Prune(version uint64) {
+	// storage pruning
+	if m.storageOpts.Interval > 0 && version > m.storageOpts.KeepRecent && version%m.storageOpts.Interval == 0 {
+		pruneVersion := version - m.storageOpts.KeepRecent - 1
+		m.pruneStorage(pruneVersion)
 	}
-	m.isStarted = false
 
-	if !m.storageOpts.Sync {
-		<-m.chStorage
-		close(m.chStorage)
-	}
-	if !m.commitmentOpts.Sync {
-		<-m.chCommitment
-		close(m.chCommitment)
+	// commitment pruning
+	if m.commitmentOpts.Interval > 0 && version > m.commitmentOpts.KeepRecent && version%m.commitmentOpts.Interval == 0 {
+		pruneVersion := version - m.commitmentOpts.KeepRecent - 1
+		m.pruneCommitment(pruneVersion)
 	}
 }
 
-// Prune prunes the state storage and state commitment.
-// It will check the pruning conditions and prune if necessary.
-func (m *Manager) Prune(height uint64) {
-	m.mtx.Lock()
-	defer m.mtx.Unlock()
-
-	if !m.isStarted {
+func (m *Manager) pruneStorage(version uint64) {
+	currentVersion := m.storageVersion.Load()
+	if 0 < currentVersion {
+		// pruning is already in progress
+		m.logger.Info("pruning storage already in progress", "version", currentVersion)
 		return
 	}
 
-	// storage pruning
-	if m.storageOpts.Interval > 0 && height > m.storageOpts.KeepRecent && height%m.storageOpts.Interval == 0 {
-		pruneHeight := height - m.storageOpts.KeepRecent - 1
-		if m.storageOpts.Sync {
-			m.pruneStorage(pruneHeight)
-		} else {
-			// it will not block if the previous pruning is still running
-			select {
-			case _, stillOpen := <-m.chStorage:
-				if stillOpen {
-					go func() {
-						m.pruneStorage(pruneHeight)
-						m.chStorage <- struct{}{}
-					}()
-				}
+	m.storageVersion.Store(version)
+	m.logger.Debug("pruning storage store", "version", version)
 
-			default:
-				m.logger.Debug("storage pruning is still running; skipping", "version", pruneHeight)
-			}
+	go func() {
```

Comment:
It doesn't make sense to me for the pruning manger to fire and forget prune in a background go thread. Shouldn't the db implementation be making the decisions about whether concurrent pruning (delete) is safe with concurrent read and write (block commit)?

Reply:
yes, the db backend is responsible for the concurrent deleting and writing.

Reply:
actually, we will create next PRs to refactor the db backend imp

Reply:
at this point I'm even wondering if having a pruning manager makes sense if this is DB dependent. 

If how pruning can be done is an implementation detail of the DB then the DB should manage this itself, no?  Maybe getting pruning options is part of the construction of the DB and the DB decides how to handle it.

Reply:
yes, it is a light manager, still feeling it is necessary 
- firstly, need the interface to interact with server cmd
- secondly, avoid the code duplication to check the pruning condition